File: ombharatiya/week1/day1/trapping-rain-water.cpp
// https://leetcode.com/problems/trapping-rain-water/

// Using DP - Left Max and Right Max Arrays 
// Time -- O(N)  -- double pass
// Space -- O(N)  


class Solution {
public:
    // A is heights array
    int trap(vector<int>& A) {
        int n = A.size();
        if (n<2) return 0;
        vector<int> right(n), left(n);
        right[n-1] = A[n-1];
        left[0] = A[0];
        for(int i=1;i<n-1;i++) {
            left[i] = max(left[i-1], A[i]);
            right[n-1-i] = max(right[n-i], A[n-1-i]);
        }
        left[n-1] = max(left[n-2], A[n-1]);
        right[0] = max(right[1], A[0]);
        
        long long sum = 0l;
        for(int i=0;i<n;i++) {
            int waterLevel = min(left[i], right[i]);
            if (waterLevel>A[i]) sum+=waterLevel-A[i];
        }
        return sum;
    }
};

// Using Stack Approach
// Time -- O(N)  -- single pass
// Space -- O(N)  

class Solution2 {
public:
    // A is heights array
    int trap(vector<int>& A) {
        int n = A.size();
        if (n<2) return 0;
        stack<int> st;
        st.push(0);
        long long sum = 0l;
        
        for(int i=1;i<n;i++) {
            while(!st.empty() && A[st.top()]<A[i]) {
                int middleBarIndex = st.top(); st.pop();
                if (!st.empty()) {
                    int boundedHeight = min(A[i], A[st.top()]) - A[middleBarIndex];
                    sum += boundedHeight * (i - st.top() - 1);
                }
            }
            st.push(i);
        }
        return sum;
    }
};

File: parnamondal/Week - 1/Day-1/Trapping Rain Water.py
# Time Complexity - O(N)
# Space Complexity - O(1)

def find (n,a):
    left=[0]*n
    right=[0]*n
    
    # for left traversing of the array
    left[0]=a[0]
    for i in range (1,n):
        left[i]=max(left[i-1],a[i])
    # for right traversing 
    right[n-1]=a[n-1]
    for i in range (n-2,-1,-1):
        right[i]=max(right[i+1],a[i])
    count=0
    
    # traversing to find the amount of trapped water 
    for i in range (n):
        count+=min(left[i],right[i])-a[i]
    return count
        
        
            
for _ in range (int(input())):
    n=int(input())
    a=list(map(int,input().split()))
    print(find(n,a))